src/engine/strat_engine/backstore/crypt.rs

```diff
@@ -0,0 +1,274 @@
+pub fn device_is_luks2(physical_path: &Path) -> Result<bool> {
+    let mut crypt_device = CryptInit::init(physical_path)?;
+    Ok(crypt_device.format_handle().get_type()? == EncryptionFormat::Luks2)
+}
+
+fn read_key(key_description: &str) -> Result<SafeMemHandle> {
+    // Attach persistent keyring to process keyring
+    match unsafe {
+        syscall(
+            SYS_keyctl,
+            KEYCTL_GET_PERSISTENT,
+            0,
+            KEY_SPEC_PROCESS_KEYRING,
+        )
+    } {
+        i if i < 0 => return Err(LibcryptErr::IOError(io::Error::last_os_error())),
+        _ => (),
+    };
+
+    let key_type_cstring = CString::new("user").expect("String is valid");
+    let key_description_cstring = CString::new(key_description).map_err(LibcryptErr::NullError)?;
+
+    // Request key ID from persistent keyring
+    let key_id = match unsafe {
+        syscall(
+            SYS_request_key,
+            key_type_cstring.as_ptr(),
+            key_description_cstring.as_ptr(),
+            std::ptr::null::<c_void>(),
+            0,
+        )
+    } {
+        i if i < 0 => return Err(LibcryptErr::IOError(io::Error::last_os_error())),
+        i => i,
+    };
+
+    let mut key_buffer = SafeMemHandle::alloc(STRATIS_KEY_SIZE)?;
+    let mut_ref = key_buffer.as_mut();
+
+    // Read key from keyring
+    match unsafe { syscall(
+        SYS_keyctl,
+        KEYCTL_READ,
+        key_id,
+        mut_ref.as_mut_ptr(),
+        mut_ref.len(),
+    ) } {
+        i if i < 0 => return Err(LibcryptErr::IOError(io::Error::last_os_error())),
+        _ => (),
+    };
+
+    Ok(key_buffer)
+}
+
+/// Lay down properly configured LUKS2 metadata on a new physical device
+pub fn initialize_encrypted_stratis_device(
+    physical_path: &Path,
+    key_description: &str,
+) -> Result<()> {
+    let mut crypt_device = CryptInit::init(physical_path)?;
+
+    crypt_device.context_handle().format::<()>(
+        EncryptionFormat::Luks2,
+        ("aes", "xts-plain64"),
+        None,
+        libcryptsetup_rs::Either::Right(STRATIS_KEY_SIZE),
+        None,
+    )?;
+
+    let key = read_key(key_description)?;
+
+    crypt_device.keyslot_handle(None).add_by_key(
+        None,
+        key.as_ref(),
+        CryptVolumeKeyFlags::empty(),
+    )?;
+
+    Ok(())
+}
+
+/// Activate encrypted Stratis device using the name stored in the
+/// Stratis token
+pub fn activate_encrypted_stratis_device(physical_path: &Path) -> Result<PathBuf> {
+    let mut crypt_device = CryptInit::init(physical_path)?;
+    crypt_device
+        .context_handle()
+        .load::<()>(EncryptionFormat::Luks2, None)?;
+
+    let stratis_device_name = get_stratis_device_name(&mut crypt_device)?;
+    crypt_device.token_handle().activate_by_token::<()>(
+        Some(stratis_device_name.as_str()),
+        Some(LUKS2_TOKEN_ID),
+        None,
+        CryptActivateFlags::empty(),
+    )?;
+
+    // Assert that physical device path is equal to the path of the provided
```

Comment:
In the final version, this should return an error, because whatever bad things happen, we don't really want to crash here. But it is definitely a reasonable thing to verify, i.e., we shouldn't go on if it is false.

Reply:
Agreed, I intend to clean that up when I submit this for review.